src/evict/evict_page.c

```diff
@@ -680,7 +681,8 @@ __evict_review(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t evict_flags, bool
              * started between previous check and acquiring snapshot. If there is a checkpoint
              * running, release the snapshot and fallback to global visibility checks.
              */
-            if (conn->txn_global.checkpoint_running) {
+            checkpoint_running = conn->txn_global.checkpoint_running;
```

Comment:
I don't think we need a lock, but Luke's comment seems on point. The comment here says "A new checkpoint might have started between previous check and acquiring snapshot", which indicates that this code really wants to read the global variable again, not use a stashed local copy. I believe it won't need any explicit controls, since bumping the snapshot should already have a barrier in it (though that's worth checking).

Reply:
What semantic do we need here exactly? Do we need to know if a checkpoint was running when the snapshot was allocated for this reconciliation? Would it be OK to know if there was a checkpoint running immediately before or immediately after?

If a checkpoint started and finished between that window, could this reconciliation possibly rewrite content in the history store?

Reply:
If the checkpoint finished in the middle, the history store should have been already checkpointed correctly. So that case should be OK. Here we want to know if there is a checkpoint on going before we get the eviction snapshot. If there is one, we give up getting the snapshot. Then we check again after we have got the snapshot, if there is one started concurrently, we release the snapshot to go for the conservative approach. If we don't detect a checkpoint, it either means there is no one started concurrently, which is OK, or one have started and finished already, which is also OK. So it is safe to use the eviction snapshot.

Reply:
```
/* Walk the array of concurrent transactions. */
    WT_ORDERED_READ(session_cnt, conn->session_cnt);
    WT_STAT_CONN_INCR(session, txn_walk_sessions);
    for (i = 0, s = txn_global->txn_shared_list; i < session_cnt; i++, s++) {
        WT_STAT_CONN_INCR(session, txn_sessions_walked);
        /*
         * Build our snapshot of any concurrent transaction IDs.
         *
         * Ignore:
         *  - Our own ID: we always read our own updates.
         *  - The ID if it is older than the oldest ID we saw. This
         *    can happen if we race with a thread that is allocating
         *    an ID -- the ID will not be used because the thread will
         *    keep spinning until it gets a valid one.
         *  - The ID if it is higher than the current ID we saw. This
         *    can happen if the transaction is already finished. In
         *    this case, we ignore this transaction because it would
         *    not be visible to the current snapshot.
         */
        while (s != txn_shared && (id = s->id) != WT_TXN_NONE && WT_TXNID_LE(prev_oldest_id, id) &&
          WT_TXNID_LT(id, current_id)) {
            /*
             * If the transaction is still allocating its ID, then we spin here until it gets its
             * valid ID.
             */
            WT_READ_BARRIER();
            if (!s->is_allocating) {
                /*
                 * There is still a chance that fetched ID is not valid after ID allocation, so we
                 * check again here. The read of transaction ID should be carefully ordered: we want
                 * to re-read ID from transaction state after this transaction completes ID
                 * allocation.
                 */
                WT_READ_BARRIER();
                if (id == s->id) {
                    txn->snapshot[n++] = id;
                    if (WT_TXNID_LT(id, pinned_id))
                        pinned_id = id;
                    break;
                }
            }
            WT_PAUSE();
        }
    }
```
There are barriers in ____txn_get_snapshot_int, but I'm confused how that relates to this issue? I think setting a volatile bool value should always be atomic.

Reply:
As I understand it an operation using a volatile isn't guaranteed to be atomic, but the read itself is guaranteed to re-read the memory location of the data instead of using a cached version.

Reply:
Yes, but assigning a value to a 64 bit memory location should be atomic on a 64 bit machine?

Reply:
I think in a multi-threaded context the answer is no? But I'm a bit unsure at this point.

Reply:
I think you need to revisit the fact that barriers are not for ensuring atomicity - they are there to ensure read and writes are not shuffled around either by the compiler or the processor hardware (which is designed to execute instructions out of order). Read barriers ensures that all previous read instructions have executed before executing the next instruction.

Reply:
`volatile` is used to indicate that variable value can change in some other thread so don't cache the copy in internal register. For example:
`for(;wait = true;)`
`   printf("Waiting ...\n"); 
`   sleep(10);`
In the example above, if we are declaring the `wait` variable with `volatile` keyword, the compiler will cache the local copy of `wait` and we will keep looping independently although some other thread might have set the `wait` to false.

Reply:
I added the comment here indicating that we have taken the decision to use snapshot because the `checkpoint_running` was false at time. However, sometime between then and now, the variable `checkpoint_running` is set to true and we have to revert whatever preparation we have done.

Reply:
Yes, I agree with Hasseb. Here we only have one state so we should not need any barrier. In addition, read write of any memory data under 64 bit should be atomic on a 64 bit machine.

Reply:
I still believe we need a barrier here however when I made the original comment I didn't check for one in the bump snapshot code, which should suffice.

Reply:
My opinion here is that we need to have a barrier somewhere, which might as well be in bump snapshot code. But my only concern is that our decision of not using a barrier is based on a function code that may chance in future.

Reply:
I think if we need a barrier, it should be a write barrier when we set the global checkpoint_running flag not when we read it.

Reply:
I'm not sure we have a shared understanding of barriers. Could we have a chat at 4?